keps/sig-network/1669-proxy-terminating-endpoints/README.md

```diff
@@ -0,0 +1,378 @@
+future testing may change this decision. See KEP 1672 for more details.
+
+## Design Details
+
+### Additions to EndpointSlice
+
+This work depends on the `Terminating` condition existing on the EndpointSlice API (see KEP 1672) in order to check the termination state of an endpoint.
+
+### kube-proxy
+
+Updates to kube-proxy when watching EndpointSlice:
+* update kube-proxy endpoints info to track terminating endpoints based on endpoint.condition.terminating in EndpointSlice.
+* update kube-proxy endpoints info to track endpoint readiness based on endpoint.condition.ready in EndpointSlice
+* within the scope of the traffic policy for a Service, iterate the following set of endpoints, picking the first set that has at least 1 ready endpoint:
+  * ready endpoints that are not terminating
+  * ready endpoints that are terminating
+
+In addition, kube-proxy's node port health check should fail if there are only `Terminating` endpoints, regardless of their readiness in order to:
+* remove the node from a loadbalancer's node pool as quickly as possible
+* gracefully handle any new connections that arrive before the loadbalancer is able to remove the node
+* allow existing connections to gracefully terminate
+
+### Test Plan
+
+#### Unit Tests
+
+kube-proxy unit tests:
+
+* Unit tests will validate the correct behavior when there are only local terminating endpoints.
+* Unit tests will validate the changein behavior against the matrix of possible Service configurations using both internalTrafficPolicy and externalTrafficPolicy.
+* Existing unit tests will validate that terminating endpoints are only used when there are no ready endpoints, otherwise ready && !terminating endpoints are used.
+* Unit tests will validate health check node port succeeds only when there are ready && !terminating endpoints.
+
+#### E2E Tests
+
+E2E tests will be added to validate that no traffic is dropped during a rolling update for a Service. E2E tests should cover all permutations of externalTrafficPolicy
+and internalTrafficPolicy.
+
+All existing E2E tests for Services should continue to pass.
+
+### Graduation Criteria
+
+#### Alpha
+
+* kube-proxy internally tracks the `terminating` and `serving` condition from EndpointSlice
+* kube-proxy falls back to terminating endpoints if and only if they are the only available endpoints.
+* feature is only enabled if the feature gate `ProxyTerminatingEndpoints` is on.
+* unit tests in kube-proxy.
+
+#### Beta
+
+* E2E tests are in place, exercising all permutations of internalTrafficPolicy and externalTrafficPolicy.
+* Metrics to publish how many Services/Endpoints are routing traffic to terminating endpoints.
+
+### Upgrade / Downgrade Strategy
+
+Behavioral changes to terminating endpoints will apply when the feature gate is enabled. It is required that the cluster has the EndpointSlice API enabled and
```

Comment:
> EndpointSlice API enabled

this is GA now, so it is not a requirement now

Reply:
Probably doesn't hurt to leave this for historical reasons (back when EndpointSlice was beta and you could turn it off)